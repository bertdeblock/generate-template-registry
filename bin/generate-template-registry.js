#!/usr/bin/env node

// eslint-disable-next-line n/no-missing-import
import { generateTemplateRegistry } from "../dist/generate-template-registry.js";

generateTemplateRegistry();
