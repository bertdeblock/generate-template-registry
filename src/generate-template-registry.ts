import chalk from "chalk";
import { camelCase, pascalCase } from "change-case";
import { pathExists, readJson } from "fs-extra/esm";
import { writeFile } from "node:fs/promises";
import { join, parse } from "node:path";
import { cwd as processCwd } from "node:process";
import recursiveReadDir from "recursive-readdir";

export async function generateTemplateRegistry(cwd = processCwd()) {
  let packageJson: any;

  try {
    packageJson = await readJson(join(cwd, "package.json"));
  } catch {
    throw new Error(
      `No "package.json" file found in the current working directory.`,
    );
  }

  const srcDir = isAddon(packageJson)
    ? isV2Addon(packageJson)
      ? "src"
      : "addon"
    : "app";

  const importRoot = isV2Addon(packageJson) ? "." : packageJson.name;

  const components = await getEntries(join(cwd, srcDir, "components"));
  const helpers = await getEntries(join(cwd, srcDir, "helpers"));
  const modifiers = await getEntries(join(cwd, srcDir, "modifiers"));

  let templateRegistryContent = "";

  if (components.length > 0) {
    const componentImports = components.map((name) => {
      return `import type ${pascalCase(name)}Component from "${importRoot}/components/${name}";`;
    });

    templateRegistryContent += "// Components:\n";
    templateRegistryContent += componentImports.join("\n") + "\n\n";
  }

  if (helpers.length > 0) {
    const helperImports = helpers.map((name) => {
      return `import type ${camelCase(name)}Helper from "${importRoot}/helpers/${name}";`;
    });

    templateRegistryContent += "// Helpers:\n";
    templateRegistryContent += helperImports.join("\n") + "\n\n";
  }

  if (modifiers.length > 0) {
    const modifierImports = modifiers.map((name) => {
      return `import type ${camelCase(name)}Modifier from "${importRoot}/modifiers/${name}";`;
    });

    templateRegistryContent += "// Modifiers:\n";
    templateRegistryContent += modifierImports.join("\n") + "\n\n";
  }

  templateRegistryContent += "export default interface Registry {\n";

  let entriesContent: string[] = [];

  if (components.length > 0) {
    let componentsContent = "  // Components:\n";

    components.forEach((name) => {
      componentsContent += `  ${angleBracketNotation(name)}: typeof ${pascalCase(name)}Component;\n`;
      componentsContent += `  "${name}": typeof ${pascalCase(name)}Component;\n`;
    });

    entriesContent.push(componentsContent);
  }

  if (helpers.length > 0) {
    let helpersContent = "  // Helpers:\n";

    helpers.forEach((name) => {
      helpersContent += `  "${name}": typeof ${camelCase(name)}Helper;\n`;
    });

    entriesContent.push(helpersContent);
  }

  if (modifiers.length > 0) {
    let modifiersContent = "  // Modifiers:\n";

    modifiers.forEach((name) => {
      modifiersContent += `  "${name}": typeof ${camelCase(name)}Modifier;\n`;
    });

    entriesContent.push(modifiersContent);
  }

  templateRegistryContent += entriesContent.join("\n") + "}\n";

  const templateRegistryPath = join(cwd, srcDir, "template-registry.ts");

  await writeFile(templateRegistryPath, templateRegistryContent);

  console.log(
    chalk.green(`Template registry generated at ${templateRegistryPath}`),
  );
}

function isAddon(packageFile: any): boolean {
  return packageFile.keywords?.includes("ember-addon");
}

function isV2Addon(packageFile: any): boolean {
  return packageFile["ember-addon"]?.version === 2;
}

async function getEntries(dir: string): Promise<string[]> {
  const paths: string[] = [];

  if (await pathExists(dir)) {
    const files = await recursiveReadDir(dir);

    files.forEach((file) => {
      const parsed = parse(file);
      const folder = parsed.dir.replace(`${dir}`, "");

      let path: string;
      if (parsed.name === "index") {
        path = folder.substring(1);
      } else {
        path = `${folder}/${parsed.name}`.substring(1);
      }

      if (paths.includes(path) === false) {
        paths.push(path);
      }
    });
  }

  return paths;
}

function angleBracketNotation(name: string) {
  const sep = "::";
  const result = name
    .split("/")
    .map((part) => pascalCase(part))
    .join(sep);

  if (result.includes(sep)) {
    return `"${result}"`;
  }

  return result;
}
