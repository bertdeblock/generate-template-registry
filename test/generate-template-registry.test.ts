import fsExtra from "fs-extra";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import recursiveCopy from "recursive-copy";
import { it } from "vitest";
import { generateTemplateRegistry } from "../src/generate-template-registry.ts";

it("generates a template registry for an app", async (ctx) => {
  const cwd = await copyBlueprint("app");
  await generateTemplateRegistry(cwd);

  const templateRegistryContent = await readFile(
    join(cwd, "app/template-registry.ts"),
    "utf-8",
  );

  ctx.expect(templateRegistryContent).toMatchSnapshot();
});

it("generates a template registry for a v1 addon", async (ctx) => {
  const cwd = await copyBlueprint("v1-addon");
  await generateTemplateRegistry(cwd);

  const templateRegistryContent = await readFile(
    join(cwd, "addon/template-registry.ts"),
    "utf-8",
  );

  ctx.expect(templateRegistryContent).toMatchSnapshot();
});

it("generates a template registry for a v2 addon", async (ctx) => {
  const cwd = await copyBlueprint("v2-addon");
  await generateTemplateRegistry(cwd);

  const templateRegistryContent = await readFile(
    join(cwd, "src/template-registry.ts"),
    "utf-8",
  );

  ctx.expect(templateRegistryContent).toMatchSnapshot();
});

async function copyBlueprint(name) {
  const cwd = join("test/output", name);

  await fsExtra.remove(cwd);
  await recursiveCopy(join("test", "blueprints", name), cwd);

  return cwd;
}
