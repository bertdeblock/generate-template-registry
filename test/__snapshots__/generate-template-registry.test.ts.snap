// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a template registry for a v1 addon 1`] = `
"// Components:
import type FooComponent from "v1-addon/components/foo";
import type BarComponent from "v1-addon/components/bar";
import type FooBarComponent from "v1-addon/components/foo/bar";

export default interface Registry {
  // Components:
  Foo: typeof FooComponent;
  "foo": typeof FooComponent;
  Bar: typeof BarComponent;
  "bar": typeof BarComponent;
  "Foo::Bar": typeof FooBarComponent;
  "foo/bar": typeof FooBarComponent;
}
"
`;

exports[`generates a template registry for a v2 addon 1`] = `
"// Components:
import type FooComponent from "./components/foo";
import type BarComponent from "./components/bar";
import type FooBarComponent from "./components/foo/bar";

export default interface Registry {
  // Components:
  Foo: typeof FooComponent;
  "foo": typeof FooComponent;
  Bar: typeof BarComponent;
  "bar": typeof BarComponent;
  "Foo::Bar": typeof FooBarComponent;
  "foo/bar": typeof FooBarComponent;
}
"
`;

exports[`generates a template registry for an app 1`] = `
"// Components:
import type FooComponent from "app/components/foo";
import type BarComponent from "app/components/bar";
import type FooBarComponent from "app/components/foo/bar";

export default interface Registry {
  // Components:
  Foo: typeof FooComponent;
  "foo": typeof FooComponent;
  Bar: typeof BarComponent;
  "bar": typeof BarComponent;
  "Foo::Bar": typeof FooBarComponent;
  "foo/bar": typeof FooBarComponent;
}
"
`;
